src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -330,144 +358,151 @@ internal override T ReadAsPropertyNameCore(ref Utf8JsonReader reader, Type typeT
+        {
+            if (namingPolicy is null)
+            {
+                return value;
+            }
 
-                        writer.WritePropertyName(formatted);
+            string converted;
+            if (!value.Contains(ValueSeparator))
+            {
+                converted = namingPolicy.ConvertName(value);
+            }
+            else
+            {
+                string[] enumValues = SplitFlagsEnum(value);
 
-                        _nameCache.TryAdd(key, formatted);
-                    }
-                    else
+                for (int i = 0; i < enumValues.Length; i++)
+                {
+                    string name = namingPolicy.ConvertName(enumValues[i]);
+                    if (name == null)
                     {
-                        // We also do not create a JsonEncodedText instance here because passing the string
-                        // directly to the writer is cheaper than creating one and not caching it for reuse.
-                        writer.WritePropertyName(original);
+                        ThrowHelper.ThrowInvalidOperationException_NamingPolicyReturnNull(namingPolicy);
                     }
-
-                    return;
+                    enumValues[i] = name;
                 }
-            }
 
-            switch (s_enumTypeCode)
-            {
-                case TypeCode.Int32:
-                    writer.WritePropertyName(Unsafe.As<T, int>(ref value));
-                    break;
-                case TypeCode.UInt32:
-                    writer.WritePropertyName(Unsafe.As<T, uint>(ref value));
-                    break;
-                case TypeCode.UInt64:
-                    writer.WritePropertyName(Unsafe.As<T, ulong>(ref value));
-                    break;
-                case TypeCode.Int64:
-                    writer.WritePropertyName(Unsafe.As<T, long>(ref value));
-                    break;
-                case TypeCode.Int16:
-                    writer.WritePropertyName(Unsafe.As<T, short>(ref value));
-                    break;
-                case TypeCode.UInt16:
-                    writer.WritePropertyName(Unsafe.As<T, ushort>(ref value));
-                    break;
-                case TypeCode.Byte:
-                    writer.WritePropertyName(Unsafe.As<T, byte>(ref value));
-                    break;
-                case TypeCode.SByte:
-                    writer.WritePropertyName(Unsafe.As<T, sbyte>(ref value));
-                    break;
-                default:
-                    ThrowHelper.ThrowJsonException();
-                    break;
+                converted = string.Join(ValueSeparator, enumValues);
             }
+
+            return converted;
+        }
+
+        private static string[] SplitFlagsEnum(string value)
+        {
+            // todo: optimize implementation here by leveraging https://github.com/dotnet/runtime/issues/934.
```

Comment:
consider adding `[ActiveIssue("https://github.com/dotnet/runtime/issues/934")]` in the comment for easier grepping